src/client.ts

```diff
@@ -569,6 +600,102 @@ in an environmental variable named FAUNA_SECRET or pass it to the Client\
   }
 }
 
+export type StreamEventHandler = (event: StreamEvent) => void;
+
+export class StreamClient {
+  #callbacks: Record<StreamEventType, StreamEventHandler[]> = {
+    start: [],
+    add: [],
+    remove: [],
+    update: [],
+    error: [],
+  };
+  #query: StreamToken;
+  #clientConfiguration: Record<string, any>;
+  #httpStreamClient: HTTPStreamClient;
+  #streamAdapter: StreamAdapter | null = null;
+
+  // TODO: make clientConfiguration and httpStreamClient optional
+  constructor(
+    query: StreamToken,
+    clientConfiguration: StreamClientConfiguration,
+    httpStreamClient: HTTPStreamClient
+  ) {
+    this.#query = query;
+    this.#clientConfiguration = clientConfiguration;
+    this.#httpStreamClient = httpStreamClient;
+  }
+
+  on(type: StreamEventType, callback: StreamEventHandler) {
+    this.#callbacks[type].push(callback);
+    return this;
+  }
+
+  start() {
+    const headers = {
+      Authorization: `Bearer ${this.#clientConfiguration.secret}`,
+    };
+
+    const streamAdapter = this.#httpStreamClient.stream({
+      data: this.#query.token,
+      headers,
+      method: "POST",
+    });
+
+    this.#streamAdapter = streamAdapter;
+
+    const handleEvents = async () => {
+      for await (const event of streamAdapter.read) {
+        // stream events are always tagged
+        const deserializedEvent: StreamEvent = TaggedTypeFormat.decode(event, {
+          long_type: this.#clientConfiguration.long_type,
+        });
+        this.#callbacks[deserializedEvent.type].forEach((callback) =>
+          callback(deserializedEvent)
+        );
+      }
+    };
+
+    handleEvents().catch((error) => {
+      throw new NetworkError("Error reading from stream", { cause: error });
+    });
+  }
+
+  async *iter(): AsyncGenerator<StreamEvent> {
+    const headers = {
+      Authorization: `Bearer ${this.#clientConfiguration.secret}`,
+    };
+
+    const streamAdapter = this.#httpStreamClient.stream({
+      data: this.#query.token,
+      headers,
+      method: "POST",
+    });
+
+    this.#streamAdapter = streamAdapter;
+
+    for await (const event of streamAdapter.read) {
+      // stream events are always tagged
+      const deserializedEvent: StreamEvent = TaggedTypeFormat.decode(event, {
+        long_type: this.#clientConfiguration.long_type,
+      });
+
+      // TODO: handle callbacks when using the async generator?
+      // this.#callbacks[deserializedEvent.type].forEach((callback) =>
+      //   callback(deserializedEvent)
+      // );
```

Comment:
Yes, definitely deserves more discussion.

If we drop the async iterator API, it will make the http2 implementation much easier. the http2 module uses callbacks itself, and that is proving to be a pain to use those to provide an iterator interface.

Reply:
personal opinion here, but my preference would be to just go with the callback api for now.  That callback pattern feels fairly common in js.  I agree that having one method  of doing it seems like a simpler and more straight forward  starting point.  
Could be good to get some input from some of th dx folks who have a bit more experience on the javascript side.

Reply:
my vote is currently on async iterator if we only go with one interface. I think it eliminates a lot of subjective questions about how to register/de-register callbacks and how to handle thrown network errors.

Reply:
I confirmed that network errors are not caught when using the callback API. We would need to pipe any errors into the error handler.

Reply:
Any errors are now caught and handled by an optional handler.